src/coreclr/jit/emit.cpp

```diff
@@ -4102,6 +4122,315 @@ void emitter::emitDispCommentForHandle(size_t handle, GenTreeFlags flag)
+                                       "be determined if the previous instruction was a call\n",
+                                       dspPtr(jmp), jmp->idDebugOnlyInfo()->idNum, group->igNum, targetGroup->igNum,
+                                       emitLabelString(targetGroup));
+                            }
+                        }
+#endif // DEBUG
+                    }
+
+                    assert(instructionDescriptor != nullptr);
+                    // instructionDescriptor is now the last instruction in the group
+
+                    if (checkedPreviousInstruction && instructionDescriptor->idIns() == jmp->idIns() &&
+                        jmp == instructionDescriptor)
+                    {
+                        CLANG_FORMAT_COMMENT_ANCHOR
+                        // the last instruction in the group is the jmp we're looking for
+                        // and it jumps to the next instruction group so we don't really need it
+
+                        if (!previousInstructionWasCall)
+                        {
+#ifdef DEBUG
+                            if (EMITVERBOSE)
+                            {
+                                printf("Removing unconditional jump [%08X/%03u] from the last instruction in "
+                                       "IG%02u to the next IG IG%02u label %s\n",
+                                       dspPtr(jmp), jmp->idDebugOnlyInfo()->idNum, group->igNum, targetGroup->igNum,
+                                       emitLabelString(targetGroup));
+                            }
+#endif
+                            UNATIVE_OFFSET sizeRemoved = instructionDescriptor->idCodeSize();
+
+                            // set state needed at the end of the loop in ADJUST_GROUP
+                            adjIG      = sizeRemoved;
+                            lstIG      = group;
+                            jmpRemoved = true;
+
+                            // unlink the jump
+                            if (previousJmp != nullptr)
+                            {
+                                previousJmp->idjNext = jmp->idjNext;
+                                if (jmp == emitJumpLast)
+                                {
+                                    emitJumpLast = previousJmp;
+                                }
+                            }
+                            else
+                            {
+                                assert(jmp == emitJumpList);
+                                emitJumpList = jmp->idjNext;
+                            }
+                            jmp->idjNext = nullptr;
+
+#if DEBUG
+                            // clear the instruction data
+                            memset((BYTE*)instructionDescriptor, 0, instructionDescriptor->idCodeSize());
+#endif
+
+                            // remove the instruction from the group
+                            group->igInsCnt -= 1;
```

Comment:
This is little dangerous because we are not sure if surrounding code would work error-free. Without updating the `igInsCnt`, you can zero the `CodeSize()` of `instructionDescriptor` and during outputting, don't output it if size is 0.

Reply:
The surrounding code seems to work ok. I ran all the tests using a checked build of the runtime and libs to make sure i hit any applicable asserts. If I change the emitter to avoid 0 length instructions the instruction counts will be wrong in dumps which will be confusing and it will then be possible to accidentally have 0 length instructions which at any other time would be an error.

Removing an instruction from the middle of a group would be much more difficult than this because any later instructions would need to be moved back up which would invalidate references to them. Specifically removing only the very last instruction is relatively safe because it can only affect the references to that value, and as you see I made sure to zero it in checked builds so that if anything touches that instruction once it has been removed it will get invalid data and usually hit an assert.

I don't quite understand the reluctance to remove the instruction rather than zero it. Other than the fact that the jit doesn't do this anywhere else (it doesn't zero size anywhere else either) is there particular reason for concern about doing it this way other than it being new?

Reply:
>  the instruction counts will be wrong in dumps which will be confusing

They are wrong currently because of `align` instructions as well. Although it will be good to fix it at one point, but I didn't notice it until I tried it today to double check. So, I wouldn't do this to just make dumps consistent. You can always output data in disasm to showcase what is going on. E.g. in below assembly code, we insert `align` instruction but then decided not to do the alignment. The "instruction count" doesn't take it out, but visually, you can see the `align` didn't happen because of `[0 bytes for IG03]` text.

<details>
<summary>40 instruction method</summary>

```asm
G_M22387_IG01:              ;; offset=0000H
       56                   push     rsi
       4883EC20             sub      rsp, 32
                                                ;; size=5 bbWeight=1    PerfScore 1.25
G_M22387_IG02:              ;; offset=0005H
       8B5114               mov      edx, dword ptr [rcx+20]
       488B01               mov      rax, gword ptr [rcx]
       3B5044               cmp      edx, dword ptr [rax+68]
       7558                 jne      SHORT G_M22387_IG09
                            align    [0 bytes for IG03]
                                                ;; size=11 bbWeight=1    PerfScore 8.00
G_M22387_IG03:              ;; offset=0010H
       8B5110               mov      edx, dword ptr [rcx+16]
       488B31               mov      rsi, gword ptr [rcx]
       3B5638               cmp      edx, dword ptr [rsi+56]
       7339                 jae      SHORT G_M22387_IG07
                                                ;; size=11 bbWeight=8    PerfScore 64.00
G_M22387_IG04:              ;; offset=001BH
       488B4610             mov      rax, gword ptr [rsi+16]
       448D4201             lea      r8d, [rdx+1]
       44894110             mov      dword ptr [rcx+16], r8d
       3B5008               cmp      edx, dword ptr [rax+8]
       7343                 jae      SHORT G_M22387_IG10
       8BD2                 mov      edx, edx
       488D1452             lea      rdx, [rdx+2*rdx]
       488D54D010           lea      rdx, bword ptr [rax+8*rdx+16]
       837A0CFF             cmp      dword ptr [rdx+12], -1
       7CD3                 jl       SHORT G_M22387_IG03
                                                ;; size=34 bbWeight=4    PerfScore 53.00
G_M22387_IG05:              ;; offset=003DH
       488B12               mov      rdx, gword ptr [rdx]
       488D4908             lea      rcx, bword ptr [rcx+8]
       E800000000           call     CORINFO_HELP_CHECKED_ASSIGN_REF
       B801000000           mov      eax, 1
                                                ;; size=17 bbWeight=0.50 PerfScore 1.88
G_M22387_IG06:              ;; offset=004EH
       4883C420             add      rsp, 32
       5E                   pop      rsi
       C3                   ret
                                                ;; size=6 bbWeight=0.50 PerfScore 0.88
G_M22387_IG07:              ;; offset=0054H
       8B4638               mov      eax, dword ptr [rsi+56]
       FFC0                 inc      eax
       894110               mov      dword ptr [rcx+16], eax
       33C0                 xor      eax, eax
       48894108             mov      qword ptr [rcx+8], rax
                                                ;; size=14 bbWeight=0.50 PerfScore 2.25
G_M22387_IG08:              ;; offset=0062H
       4883C420             add      rsp, 32
       5E                   pop      rsi
       C3                   ret
                                                ;; size=6 bbWeight=0.50 PerfScore 0.88
G_M22387_IG09:              ;; offset=0068H
       FF1500000000         call     [System.ThrowHelper:ThrowInvalidOperationException_InvalidOperation_EnumFailedVersion()]
       CC                   int3
                                                ;; size=7 bbWeight=0    PerfScore 0.00
G_M22387_IG10:              ;; offset=006FH
       E800000000           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3
                                                ;; size=6 bbWeight=0    PerfScore 0.00

; Total bytes of code 117, prolog size 5, PerfScore 143.83, instruction count 40, allocated bytes for code 117 (MethodHash=d4f7a88c) for method Enumerator[__Canon,Boolean][System.__Canon,System.Boolean]:MoveNext():bool:this
; ============================================================
```
</details>

> is there particular reason for concern about doing it this way other than it being new?

The reason I am reluctant is because the notion of updating the `insCount` is new and I am not sure if all the code of all targets (when we support armarch) is resilient to this. Nor we are not sure if we have enough test coverage to all possible affected paths. Errors can come not just from asserts but also silent bad code gen which takes time to debug and investigate.

Reply:
To add on, the insGroup/instrDesc system is not well designed to support removal. I don't think this optimization justifies adding that support in this particular case; it's hard to be confident in the correctness of the result. There are various things that are pushing us towards generalizing this in the future, like introducing general late peephole optimizations.

Even for this change, if you didn't remove the instruction, you wouldn't need `genUpdateLiveRangesForTruncatedIGs`, or the emitLocation update API, right?

Reply:
> Even for this change, if you didn't remove the instruction, you wouldn't need genUpdateLiveRangesForTruncatedIGs, or the emitLocation update API, right?

+1 on that.

Reply:
Ok. So how would I change the instruction?

> Even for this change, if you didn't remove the instruction, you wouldn't need `genUpdateLiveRangesForTruncatedIGs`, or the emitLocation update API, right?

True. But then don't the uncounted align instructions mean there are holes in that anyway?

So how would I change an instruction? There's no prior art for that any more than there was for removing an instruction.

It might take a while to get these changes done. Unfortunately rebasing this has pulled in some change which prevents me being able to debug into the jit and I don't know how to fix it. I'm not going back to printf debugging again, it was a huge time sink, so I'm going to have to wait until someone else finds and fixes the problem.

Reply:
> this has pulled in some change which prevents me being able to debug into the jit 

What errors are you getting?

Reply:
No errors. Build is clean. When I set breakpoints in the change code they won't activate due to the ide saying the source differs from the built version. Stepping seems to be stepping into the previous version of the code. Clearly something weird is going on but I can't see where. I've debugged into EEstartup far enough to verify that the clrjit from the corerun folder is being loaded. It was working fine until I rebased. I've done a full repo clean and rebuild and re-setup and arrived at the same point so I know it's not a corruption/environment issue. I'm primarily a managed code developer so I don't have the deep knowledge of the unmanaged dev side needed to find out what's going on.

Reply:
> there are holes in that anyway?

There are holes in `instrDesc` data structure (which is fine to have) but not in emitted machine code.

> far enough to verify that the clrjit from the corerun folder is being loaded

You can also check the `Modules` window (if you are using VS for debugging) to check the path of loaded module and its symbols.

![image](https://user-images.githubusercontent.com/12488060/167963218-4fcf4456-0ad5-40e3-93ad-714b97067cf4.png)

> When I set breakpoints in the change code they won't activate 

Are you doing `Checked` or `Debug` build? Do you run from `CORE_ROOT`?

Reply:
Checked build. I use the [core clr debugging workflow](https://github.com/dotnet/runtime/blob/f44b0ba183ec2680f3d35b8615d474d1d50bc94e/docs/workflow/debugging/coreclr/debugging.md) instructions. I was working normally with it and able to debug until I rebased and then I found I wasn't able to debug new changes. This morning I did a `git clean -xdf` and built main cleanly and then switch to my branch and re-setup for debugging, the same inability to debug happened. I've been making small changes for a while now so I'm familiar enough with my workflow to get from bare repo to working reasonably well.